common/persistence/nosql/nosqlplugin/cassandra/gocql/errors.go

```diff
@@ -39,33 +41,50 @@ func ConvertError(
 	operation string,
 	err error,
 ) error {
-
-	switch err {
-	case nil:
+	if err == nil {
 		return nil
-	case context.DeadlineExceeded, gocql.ErrTimeoutNoResponse, gocql.ErrConnectionClosed:
+	}
+
+	if errors.Is(err, context.DeadlineExceeded) || errors.Is(err, gocql.ErrTimeoutNoResponse) || errors.Is(err, gocql.ErrConnectionClosed) {
 		return &persistence.TimeoutError{Msg: fmt.Sprintf("operation %v encountered %v", operation, err.Error())}
-	case gocql.ErrNotFound:
+	}
+	if errors.Is(err, gocql.ErrNotFound) {
 		return serviceerror.NewNotFound(fmt.Sprintf("operation %v encountered %v", operation, err.Error()))
 	}
 
-	switch v := err.(type) {
-	case *gocql.RequestErrWriteTimeout:
-		return &persistence.TimeoutError{Msg: fmt.Sprintf("operation %v encountered %v", operation, err.Error())}
-	case gocql.RequestError:
-		if v.Code() == gocql.ErrCodeOverloaded {
+	var cqlTimeoutErr gocql.RequestErrWriteTimeout
+	if errors.As(err, &cqlTimeoutErr) {
+		return &persistence.TimeoutError{Msg: fmt.Sprintf("operation %v encountered %v", operation, cqlTimeoutErr.Error())}
+	}
+
+	var cqlRequestErr gocql.RequestError
+	if errors.As(err, &cqlRequestErr) {
+		if cqlRequestErr.Code() == gocql.ErrCodeOverloaded {
 			return &serviceerror.ResourceExhausted{
 				Cause:   enumspb.RESOURCE_EXHAUSTED_CAUSE_SYSTEM_OVERLOADED,
 				Scope:   enumspb.RESOURCE_EXHAUSTED_SCOPE_SYSTEM,
-				Message: fmt.Sprintf("operation %v encountered %v", operation, err.Error()),
+				Message: fmt.Sprintf("operation %v encountered %v", operation, cqlRequestErr.Error()),
 			}
 		}
-		return serviceerror.NewUnavailable(fmt.Sprintf("operation %v encountered %v", operation, err.Error()))
-	default:
-		return serviceerror.NewUnavailable(fmt.Sprintf("operation %v encountered %v", operation, err.Error()))
+
+		if cqlRequestErr.Code() == gocql.ErrCodeInvalid {
+			if strings.Contains(strings.ToLower(cqlRequestErr.Message()), "disk usage exceeds failure threshold") {
+				return &serviceerror.ResourceExhausted{
+					// TODO(alextalks): correct to PERSISTENCE_STORAGE_LIMIT once https://github.com/temporalio/api/pull/417 merges.
+					Cause:   enumspb.RESOURCE_EXHAUSTED_CAUSE_PERSISTENCE_LIMIT,
+					Scope:   enumspb.RESOURCE_EXHAUSTED_SCOPE_SYSTEM,
+					Message: fmt.Sprintf("operation %v encountered %v", operation, cqlRequestErr.Error()),
+				}
+			}
+
+			return serviceerror.NewInvalidArgument(fmt.Sprintf("operation %v encountered %v", operation, cqlRequestErr.Error()))
```

Comment:
This line is redundant with the one below it

Reply:
And may not even be necessary as the final return from the function captures this